ruma-client-api/src/r0/capabilities/get_capabilities/iter.rs

```diff
@@ -0,0 +1,63 @@
+use std::{borrow::Cow, collections::btree_map::Iter};
+
+use serde_json::Value as JsonValue;
+
+use super::Capabilities;
+
+/// Reference to a capability
+#[derive(Debug)]
+pub struct CapabilityRef<'a> {
+    name: &'a str,
+    value: Option<&'a JsonValue>,
+    caps: &'a Capabilities,
+}
+
+impl<'a> CapabilityRef<'a> {
+    pub fn name(&self) -> &'a str {
+        self.name
+    }
+
+    pub fn value(&self) -> Cow<'a, JsonValue> {
+        match self.value {
+            // unknown capability from btreemap iterator
+            Some(val) => Cow::Borrowed(val),
+            // O(1) lookup of known capability
+            None => self.caps.get(self.name).unwrap(),
+        }
+    }
+}
+
+/// An iterator over capabilities
+#[derive(Debug)]
+pub struct CapabilitiesIter<'a> {
+    caps: &'a Capabilities,
+    pos: usize,
+    custom_caps_iterator: Iter<'a, String, JsonValue>,
+}
+
+impl<'a> CapabilitiesIter<'a> {
+    pub fn new(caps: &'a Capabilities) -> Self {
+        Self { caps, pos: 0, custom_caps_iterator: caps.custom_capabilities.iter() }
+    }
+}
+
+impl<'a> Iterator for CapabilitiesIter<'a> {
+    type Item = CapabilityRef<'a>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.pos {
+            0 => {
+                self.pos += 1;
```

Comment:
I think it would be a bit cleaner if these were moved to after the `match` (`let res = match ...; self.pos += 1; res`).

Reply:
The reason it's written that way is because I am not incrementing `self.pos` if there's nothing left to iterate (`self.custom_caps_iterator.next()`is `None`)
Should I incrementing`self.pos` regardless of `self.custom_caps_iterator.next()`?

Reply:
Oh, I see. It shouldn't matter in practice I guess but now I agree that your solution is better. Actually the `self.pos += 1` inside the closure in the last branch seems pointless, but also it should be impossible to have a `BTreeMap` with more than `usize` elements 😄